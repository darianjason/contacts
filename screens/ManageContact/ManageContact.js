import React, {useReducer, useState, useLayoutEffect} from 'react';
import {ScrollView, ActivityIndicator, Alert} from 'react-native';
import {useDispatch, useSelector} from 'react-redux';
import {useNavigation} from '@react-navigation/native';

import {ContactForm} from '../../components/Contacts';
import {Button, ErrorOverlay} from '../../components/ui';
import {
  addContact,
  editContact,
  removeContact,
} from '../../store/contactsSlice';
import {Colors, INPUT_UPDATE} from '../../constants';
import styles from './ManageContact.styles';

const formReducer = (state, action) => {
  if (action.type === INPUT_UPDATE) {
    const values = {
      ...state.inputValues,
      [action.input]: action.value,
    };

    const validities = {
      ...state.inputValidities,
      [action.input]: action.isValid,
    };

    const errors = {
      ...state.inputErrors,
      [action.input]: action.errorText,
    };

    let formIsValid = true;
    for (const key in validities) {
      formIsValid = formIsValid && validities[key];
    }

    return {
      formIsValid: formIsValid,
      inputValues: values,
      inputValidities: validities,
      inputErrors: errors,
    };
  }

  return state;
};

const cancelHandler = navigation => {
  navigation.goBack();
};

const saveHandler = async saveParams => {
  const {
    isEditing,
    navigation,
    formState,
    id,
    dispatch,
    setIsSaving,
    setError,
  } = saveParams;
  const {inputValues, formIsValid} = formState;
  const {firstName, lastName, age, photo} = inputValues;

  const contact = {
    firstName: firstName,
    lastName: lastName,
    age: +age,
    photo: photo ? photo : 'N/A',
  };

  if (!formIsValid) {
    Alert.alert('Invalid data', 'Please check your inputs.');
    return;
  }

  setIsSaving(true);

  try {
    if (!isEditing) {
      await dispatch(addContact(contact));
      navigation.reset({
        index: 0,
        routes: [{name: 'Contacts'}],
      });
    } else {
      await dispatch(editContact(id, contact));
      contact.id = id;
      navigation.reset({
        index: 1,
        routes: [
          {name: 'Contacts'},
          {name: 'ContactDetails', params: {contact}},
        ],
      });
    }
  } catch (error) {
    setError(error.message);
    setIsSaving(false);
  }
};

const deleteHandler = async deleteParams => {
  const {dispatch, selectedContact, navigation, setIsDeleting, setError} =
    deleteParams;
  const {id} = selectedContact;

  setIsDeleting(true);

  try {
    await dispatch(removeContact(id));
    navigation.reset({
      index: 0,
      routes: [{name: 'Contacts'}],
    });
  } catch (error) {
    setError(error.message);
    setIsDeleting(false);
  }
};

const confirmDelete = deleteParams => {
  const {selectedContact} = deleteParams;
  const {firstName, lastName} = selectedContact;

  Alert.alert(
    'Are you sure?',
    `Do you want to delete ${firstName} ${lastName}?`,
    [
      {text: 'No', style: 'default'},
      {
        text: 'Yes',
        style: 'destructive',
        onPress: () => deleteHandler(deleteParams),
      },
    ],
  );
};

const setNavOptions = navParams => {
  const {
    navigation,
    isEditing,
    formState,
    id,
    dispatch,
    isSaving,
    setIsSaving,
    setError,
  } = navParams;
  const saveParams = {
    isEditing,
    navigation,
    formState,
    id,
    dispatch,
    setIsSaving,
    setError,
  };

  const {formIsValid} = formState;

  navigation.setOptions({
    title: isEditing ? 'Edit Contact' : 'Add Contact',
    headerLeft: () => (
      <Button
        icon="times-circle"
        color={Colors.red}
        onPress={() => cancelHandler(navigation)}
        style={styles.cancelButton}
      />
    ),
    headerRight: () =>
      isSaving ? (
        <ActivityIndicator size="small" color={Colors.accent} />
      ) : (
        <Button
          icon="save"
          onPress={() => saveHandler(saveParams)}
          disabled={!formIsValid}
          style={!formIsValid && styles.disabled}
        />
      ),
  });
};

const errorHandler = setError => {
  setError(null);
};

const ManageContact = ({route}) => {
  const id = route.params?.id;
  const isEditing = !!id;

  const [isSaving, setIsSaving] = useState(false);
  const [isDeleting, setIsDeleting] = useState(false);
  const [error, setError] = useState();

  const placeholderContact = {
    id: '',
    firstName: '',
    lastName: '',
    age: '',
    photo: '',
  };

  const contacts = useSelector(state => state.contacts.contacts);
  const selectedContact = isDeleting
    ? placeholderContact
    : contacts.find(contact => contact.id === id);

  const dispatch = useDispatch();
  const navigation = useNavigation();

  const [formState, dispatchFormState] = useReducer(formReducer, {
    inputValues: {
      firstName: isEditing ? selectedContact.firstName : '',
      lastName: isEditing ? selectedContact.lastName : '',
      age: isEditing ? selectedContact.age.toString() : '',
      photo: isEditing ? selectedContact.photo : 'N/A',
    },
    inputValidities: {
      firstName: !!isEditing,
      lastName: !!isEditing,
      age: !!isEditing,
      photo: !!isEditing,
    },
    inputErrors: {
      firstName: '',
      lastName: '',
      age: '',
      photo: '',
    },
    formIsValid: !!isEditing,
  });

  useLayoutEffect(() => {
    const navParams = {
      navigation,
      isEditing,
      formState,
      id,
      dispatch,
      isSaving,
      setIsSaving,
      setError,
    };
    setNavOptions(navParams);
  }, [
    navigation,
    isEditing,
    formState,
    id,
    dispatch,
    isSaving,
    setIsSaving,
    setError,
  ]);

  const deleteParams = {
    dispatch,
    selectedContact,
    navigation,
    setIsDeleting,
    setError,
  };

  if (error && (!isDeleting || !isSaving)) {
    return (
      <ErrorOverlay message={error} onConfirm={() => errorHandler(setError)} />
    );
  }

  return (
    <ScrollView contentContainerStyle={styles.screen}>
      <ContactForm
        formState={formState}
        dispatchFormState={dispatchFormState}
        isEditing={isEditing}
      />
      {isEditing &&
        (!isDeleting ? (
          <Button
            icon="trash-alt"
            color={Colors.red}
            onPress={() => confirmDelete(deleteParams)}
          >
            Delete Contact
          </Button>
        ) : (
          <ActivityIndicator size="small" color={Colors.red} />
        ))}
    </ScrollView>
  );
};

export default ManageContact;
