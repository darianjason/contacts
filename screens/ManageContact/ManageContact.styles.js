// @flow

import {StyleSheet} from 'react-native';
import type {ViewStyleProp} from 'react-native/Libraries/StyleSheet/StyleSheet';

import {Colors} from '../../constants';

type Styles = {
  screen: ViewStyleProp,
  cancelButton: ViewStyleProp,
  disabled: ViewStyleProp,
};

const styles: Styles = StyleSheet.create({
  screen: {
    flex: 1,
    padding: 36,
    backgroundColor: Colors.accent,
  },
  cancelButton: {
    marginRight: 36,
  },
  disabled: {
    opacity: 0.5,
  },
});

export default styles;
