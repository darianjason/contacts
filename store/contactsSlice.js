import {createSlice} from '@reduxjs/toolkit';

const contactsSlice = createSlice({
  name: 'contacts',
  initialState: {
    contacts: [],
  },
  reducers: {
    setContacts: (state, action) => {
      state.contacts = action.payload;
    },
    insertContact: (state, action) => {
      state.contacts.push(action.payload);
    },
    updateContact: (state, action) => {
      const {id, data} = action.payload;
      const updatedContactIndex = state.contacts.findIndex(
        contact => contact.id === id,
      );
      state.contacts[updatedContactIndex] = data;
    },
    deleteContact: (state, action) => {
      const {id} = action.payload;
      const index = state.contacts.findIndex(contact => contact.id === id);
      state.contacts.splice(index, 1);
    },
  },
});

const {actions, reducer} = contactsSlice;
export const {setContacts, insertContact, updateContact, deleteContact} =
  actions;
export default reducer;

const BASE_URL = 'https://simple-contact-crud.herokuapp.com';

export const fetchContacts = () => async dispatch => {
  try {
    const response = await fetch(`${BASE_URL}/contact`);
    const contacts = await response.json();
    dispatch(setContacts(contacts.data));
  } catch (error) {
    throw error;
  }
};

export const addContact = contact => async dispatch => {
  try {
    const response = await fetch(`${BASE_URL}/contact`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(contact),
    });

    const responseJson = await response.json();
    console.log(responseJson);

    dispatch(insertContact(contact));
  } catch (error) {
    throw error;
  }
};

export const editContact = (id, data) => async dispatch => {
  try {
    const response = await fetch(`${BASE_URL}/contact/${id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    });

    const responseJson = await response.json();
    console.log(responseJson);

    dispatch(updateContact(id, data));
  } catch (error) {
    throw error;
  }
};

export const removeContact = id => async dispatch => {
  try {
    const response = await fetch(`${BASE_URL}/contact/${id}`, {
      method: 'DELETE',
    });

    const responseJson = await response.json();
    console.log(responseJson);

    dispatch(deleteContact(id));
  } catch (error) {
    throw error;
  }
};
