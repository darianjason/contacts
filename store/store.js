// @flow

import {configureStore} from '@reduxjs/toolkit';

import contactsReducer from './contactsSlice';

type Contact = {
  id?: string,
  firstName: string,
  lastName: string,
  age: number | string,
  photo: string,
};

type Store = {
  contacts: Array<Contact>,
};

export const store: Store = configureStore({
  reducer: {
    contacts: contactsReducer,
  },
});
