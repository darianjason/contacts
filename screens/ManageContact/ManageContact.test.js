import React from 'react';
import {render} from '@testing-library/react-native';
import {Provider} from 'react-redux';

import {store} from '../../store/store';
import ManageContact, {saveHandler, deleteHandler} from './ManageContact';

describe('Manage Contact screen', () => {
  test('saves correct contact data', async () => {
    const contact = {
      firstName: 'Hello',
      lastName: 'World',
      age: 25,
      photo: 'uri',
    };

    const isEditing = false;
    const navigation = {
      reset: jest.fn(),
    };
    const formState = {
      inputValues: {
        firstName: contact.firstName,
        lastName: contact.lastName,
        age: contact.age,
        photo: contact.photo,
      },
      inputValidities: {
        firstName: true,
        lastName: true,
        age: true,
        photo: true,
      },
      formIsValid: true,
    };
    const dispatch = jest.fn();
    const setIsSaving = jest.fn();
    const setError = jest.fn();

    const saveParams = {
      isEditing,
      navigation,
      formState,
      dispatch,
      setIsSaving,
      setError,
    };

    const response = {message: 'contact saved'};

    global.fetch = jest.fn();
    fetch.mockResolvedValueOnce({
      status: 201,
      ok: true,
      json: async () => response,
    });

    const savedContact = await saveHandler(saveParams);

    expect(savedContact).toEqual(contact);
  });

  test('deletes contact with correct ID', async () => {
    const dispatch = jest.fn();
    const selectedContact = {
      id: 'test',
      firstName: 'Hello',
      lastName: 'World',
      age: 25,
      photo: 'uri',
    };
    const navigation = {
      reset: jest.fn(),
    };
    const setIsDeleting = jest.fn();
    const setError = jest.fn();

    const deleteParams = {
      dispatch,
      selectedContact,
      navigation,
      setIsDeleting,
      setError,
    };

    const response = {message: 'contact deleted'};

    global.fetch = jest.fn();
    fetch.mockResolvedValueOnce({
      status: 202,
      ok: true,
      json: async () => response,
    });

    const deletedContactId = await deleteHandler(deleteParams);

    expect(deletedContactId).toEqual(selectedContact.id);
  });

  test('renders correctly', async () => {
    const route = {
      params: {},
    };

    const tree = render(
      <Provider store={store}>
        <ManageContact route={route} />
      </Provider>,
    );

    expect(tree).toMatchSnapshot();
  });
});
