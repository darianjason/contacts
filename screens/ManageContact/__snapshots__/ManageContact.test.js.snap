// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manage Contact screen renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "backgroundColor": "#E2EEF3",
      "flex": 1,
      "padding": 36,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "marginBottom": 16,
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginBottom": 16,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "#F7F4F3",
              "borderRadius": 8,
              "height": 166.75,
              "justifyContent": "center",
              "overflow": "hidden",
              "width": 166.75,
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              Object {
                "color": "#142B3E",
                "fontFamily": "Manrope-Bold",
                "fontSize": 24,
                "letterSpacing": 2,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "justifyContent": "center",
              "marginBottom": 16,
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#142B3E",
                "fontFamily": "Manrope-Bold",
                "letterSpacing": 0.5,
                "marginBottom": 8,
              },
              false,
            ]
          }
        >
          First Name
        </Text>
        <TextInput
          alphanumeric={true}
          autoCapitalize="words"
          autoCorrect={false}
          id="firstName"
          initialValue=""
          initiallyValid={false}
          label="First Name"
          maxLength={30}
          minLength={3}
          onBlur={[Function]}
          onChangeText={[Function]}
          onInputChange={[Function]}
          required={true}
          style={
            Array [
              Object {
                "backgroundColor": "#F7F4F3",
                "borderRadius": 4,
                "fontFamily": "Manrope-Regular",
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              false,
            ]
          }
          value=""
        />
      </View>
      <View
        style={
          Array [
            Object {
              "justifyContent": "center",
              "marginBottom": 16,
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#142B3E",
                "fontFamily": "Manrope-Bold",
                "letterSpacing": 0.5,
                "marginBottom": 8,
              },
              false,
            ]
          }
        >
          Last Name
        </Text>
        <TextInput
          alphanumeric={true}
          autoCapitalize="words"
          autoCorrect={false}
          id="lastName"
          initialValue=""
          initiallyValid={false}
          label="Last Name"
          maxLength={30}
          minLength={3}
          onBlur={[Function]}
          onChangeText={[Function]}
          onInputChange={[Function]}
          required={true}
          style={
            Array [
              Object {
                "backgroundColor": "#F7F4F3",
                "borderRadius": 4,
                "fontFamily": "Manrope-Regular",
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              false,
            ]
          }
          value=""
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "justifyContent": "center",
                "marginBottom": 16,
              },
              Object {
                "flex": 1,
                "marginRight": 24,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#142B3E",
                  "fontFamily": "Manrope-Bold",
                  "letterSpacing": 0.5,
                  "marginBottom": 8,
                },
                false,
              ]
            }
          >
            Age
          </Text>
          <TextInput
            id="age"
            initialValue=""
            initiallyValid={false}
            keyboardType="number-pad"
            label="Age"
            max={100}
            min={1}
            numeric={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onInputChange={[Function]}
            required={true}
            style={
              Array [
                Object {
                  "backgroundColor": "#F7F4F3",
                  "borderRadius": 4,
                  "fontFamily": "Manrope-Regular",
                  "paddingHorizontal": 8,
                  "paddingVertical": 4,
                },
                false,
              ]
            }
            value=""
          />
        </View>
        <View
          style={
            Array [
              Object {
                "justifyContent": "center",
                "marginBottom": 16,
              },
              Object {
                "flex": 4,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#142B3E",
                  "fontFamily": "Manrope-Bold",
                  "letterSpacing": 0.5,
                  "marginBottom": 8,
                },
                false,
              ]
            }
          >
            Photo URL
          </Text>
          <TextInput
            autoCapitalize="none"
            autoCorrect={false}
            id="photo"
            initialValue="N/A"
            initiallyValid={true}
            keyboardType="url"
            label="Photo URL"
            onBlur={[Function]}
            onChangeText={[Function]}
            onInputChange={[Function]}
            style={
              Array [
                Object {
                  "backgroundColor": "#F7F4F3",
                  "borderRadius": 4,
                  "fontFamily": "Manrope-Regular",
                  "paddingHorizontal": 8,
                  "paddingVertical": 4,
                },
                false,
              ]
            }
            value="N/A"
          />
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
